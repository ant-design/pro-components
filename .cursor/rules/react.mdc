---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
