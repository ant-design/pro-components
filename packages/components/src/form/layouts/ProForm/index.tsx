import type { FormProps } from 'antd';
import { Form } from 'antd';
import React from 'react';
import type { CommonFormProps } from '../../BaseForm';
import { BaseForm } from '../../BaseForm';
import { EditOrReadOnlyContext } from '../../BaseForm/EditOrReadOnlyContext';
import { Group, ProFormItem } from '../../components';

export type ProFormProps<
  T = Record<string, any>,
  U = Record<string, any>,
> = Omit<FormProps<T>, 'onFinish'> & CommonFormProps<T, U>;

function ProForm<T = Record<string, any>>(
  props: ProFormProps<T> & {
    children?: React.ReactNode | React.ReactNode[];
  },
) {
  return (
    <BaseForm
      layout="vertical"
      contentRender={(items, submitter) => {
        return (
          <>
            {items}
            {submitter}
          </>
        );
      }}
      {...props}
    />
  );
}

ProForm.Group = Group;
ProForm.useForm = Form.useForm;
ProForm.Item = ProFormItem;
ProForm.useWatch = Form.useWatch;
ProForm.ErrorList = Form.ErrorList;
ProForm.Provider = Form.Provider;
ProForm.useFormInstance = Form.useFormInstance;
ProForm.EditOrReadOnlyContext = EditOrReadOnlyContext;

export { ProForm };
