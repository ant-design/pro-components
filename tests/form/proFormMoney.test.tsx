import ProForm, { ProFormMoney } from '@ant-design/pro-form';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import { ConfigProvider } from 'antd';
import enGBIntl from 'antd/lib/locale/en_GB';
import { act } from 'react';

afterEach(() => {
  cleanup();
});

describe('ðŸ’µ ProFormMoney', () => {
  it('ðŸ’µ ProFormMoney value expect number', async () => {
    const fn = vi.fn();
    const { container } = render(
      <ProForm
        onFinish={async (values) => {
          fn(values.amount);
        }}
      >
        <ProFormMoney name="amount" initialValue={44.33} />
      </ProForm>,
    );

    expect(container.querySelectorAll('input#amount')[0]).toHaveAttribute(
      'value',
      'Â¥ 44.33',
    );

    act(() => {
      fireEvent.click(container.querySelector('button.ant-btn-primary')!);
    });

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith(44.33);
    });
    expect(container).toMatchSnapshot();
  });

  it('ðŸ’µ moneySymbol with global locale', async () => {
    const fn = vi.fn();
    const { container } = render(
      <ConfigProvider locale={enGBIntl}>
        <ProForm
          onFinish={async (values) => {
            fn(values.amount);
          }}
        >
          <ProFormMoney name="amount" initialValue={44.33} />
        </ProForm>
      </ConfigProvider>,
    );

    expect(container.querySelectorAll('input#amount')[0]).toHaveAttribute(
      'value',
      'Â£ 44.33',
    );

    act(() => {
      fireEvent.click(container.querySelector('button.ant-btn-primary')!);
    });

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith(44.33);
    });
    expect(container).toMatchSnapshot();
  });

  it('ðŸ’µ moneySymbol with custom locale', async () => {
    const fn = vi.fn();
    const { container } = render(
      <ProForm
        onFinish={async (values) => {
          fn(values.amount);
        }}
      >
        <ProFormMoney name="amount" initialValue={44.33} locale="en-US" />
      </ProForm>,
    );

    expect(container.querySelectorAll('input#amount')[0]).toHaveAttribute(
      'value',
      '$ 44.33',
    );

    fireEvent.click(container.querySelector('button.ant-btn-primary')!);

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith(44.33);
    });
    expect(container).toMatchSnapshot();
  });
  it('ðŸ’µ moneySymbol with custom symbol', async () => {
    const fn = vi.fn();
    const { container } = render(
      <ProForm
        onFinish={async (values) => {
          fn(values.amount);
        }}
      >
        <ProFormMoney name="amount" initialValue={44.33} customSymbol="ðŸ’°" />
      </ProForm>,
    );

    expect(container.querySelectorAll('input#amount')[0]).toHaveAttribute(
      'value',
      'ðŸ’° 44.33',
    );
    act(() => {
      fireEvent.click(container.querySelector('button.ant-btn-primary')!);
    });
    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith(44.33);
    });
    expect(container).toMatchSnapshot();
  });
  it('ðŸ’µ can not input negative', async () => {
    const fn = vi.fn();
    const { container } = render(
      <ProForm
        onFinish={async (values) => {
          fn(values.amount);
        }}
      >
        <ProFormMoney name="amount" min={0} />
      </ProForm>,
    );

    expect(container.querySelectorAll('input#amount')[0]).toHaveAttribute(
      'value',
      '',
    );

    await fireEvent.change(container.querySelector('input#amount')!, {
      target: {
        value: '-55.33',
      },
    });
    act(() => {
      fireEvent.click(container.querySelector('button.ant-btn-primary')!);
    });
    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith(undefined);
    });
    expect(container).toMatchSnapshot();
  });
  it('ðŸ’µ can input negative', async () => {
    const fn = vi.fn();
    const { container } = render(
      <ProForm
        onFinish={async (values) => {
          fn(values.amount);
        }}
      >
        <ProFormMoney name="amount" />
      </ProForm>,
    );

    expect(container.querySelectorAll('input#amount')[0]).toHaveAttribute(
      'value',
      '',
    );

    await fireEvent.change(container.querySelector('input#amount')!, {
      target: {
        value: '-55.33',
      },
    });

    expect(container.querySelectorAll('input#amount')[0]).toHaveAttribute(
      'value',
      'Â¥ -55.33',
    );

    fireEvent.click(container.querySelector('button.ant-btn-primary')!);

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith(-55.33);
    });
    expect(container).toMatchSnapshot();
  });

  it('ðŸ’µ update money precision when init', async () => {
    const fn = vi.fn();
    const { container } = render(
      <ProForm
        onFinish={async (values) => {
          fn(values.amount);
        }}
      >
        <ProFormMoney
          name="amount"
          initialValue={444444444.333333333}
          fieldProps={{ precision: 2 }}
          customSymbol="ðŸ’°"
        />
      </ProForm>,
    );

    expect(container.querySelectorAll('input#amount')[0]).toHaveAttribute(
      'value',
      'ðŸ’° 444,444,444.33',
    );

    fireEvent.click(container.querySelector('button.ant-btn-primary')!);

    await waitFor(() => {
      expect(fn).toHaveBeenCalledWith(444444444.333333333);
    });
    expect(container).toMatchSnapshot();
  });
});
