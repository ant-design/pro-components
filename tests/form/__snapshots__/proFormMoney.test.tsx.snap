// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`üíµ ProFormMoney > üíµ ProFormMoney value expect number 1`] = `
<div>
  <form
    autocomplete="off"
    class="ant-form ant-form-vertical ant-pro-form"
  >
    <input
      style="display: none;"
      type="text"
    />
    <div
      class="ant-form-item ant-form-item-vertical"
    >
      <div
        class="ant-row ant-form-item-row"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-input-number ant-input-number-in-form-item ant-input-number-outlined ant-input-number-focused"
                style="width: 100%;"
              >
                <div
                  class="ant-input-number-handler-wrap"
                >
                  <span
                    aria-disabled="false"
                    aria-label="Increase Value"
                    class="ant-input-number-handler ant-input-number-handler-up"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="up"
                      class="anticon anticon-up ant-input-number-handler-up-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    aria-disabled="false"
                    aria-label="Decrease Value"
                    class="ant-input-number-handler ant-input-number-handler-down"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-input-number-handler-down-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-input-number-input-wrap"
                >
                  <input
                    aria-describedby="test-id"
                    aria-valuenow="44.33"
                    autocomplete="off"
                    class="ant-input-number-input"
                    id="amount"
                    placeholder="ËØ∑ËæìÂÖ•"
                    role="spinbutton"
                    step="1"
                    value="¬• 44.33"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; gap: 8px; align-items: center;"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
        type="button"
      >
        <span>
          Èáç ÁΩÆ
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
        type="button"
      >
        <span>
          Êèê ‰∫§
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`üíµ ProFormMoney > üíµ can input negative 1`] = `
<div>
  <form
    autocomplete="off"
    class="ant-form ant-form-vertical ant-pro-form"
  >
    <input
      style="display: none;"
      type="text"
    />
    <div
      class="ant-form-item ant-form-item-has-success ant-form-item-vertical"
    >
      <div
        class="ant-row ant-form-item-row"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-input-number ant-input-number-in-form-item ant-input-number-outlined ant-input-number-status-success ant-input-number-focused"
                style="width: 100%;"
              >
                <div
                  class="ant-input-number-handler-wrap"
                >
                  <span
                    aria-disabled="false"
                    aria-label="Increase Value"
                    class="ant-input-number-handler ant-input-number-handler-up"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="up"
                      class="anticon anticon-up ant-input-number-handler-up-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    aria-disabled="false"
                    aria-label="Decrease Value"
                    class="ant-input-number-handler ant-input-number-handler-down"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-input-number-handler-down-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-input-number-input-wrap"
                >
                  <input
                    aria-describedby="test-id"
                    aria-valuenow="-55.33"
                    autocomplete="off"
                    class="ant-input-number-input"
                    id="amount"
                    placeholder="ËØ∑ËæìÂÖ•"
                    role="spinbutton"
                    step="1"
                    value="¬• -55.33"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; gap: 8px; align-items: center;"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
        type="button"
      >
        <span>
          Èáç ÁΩÆ
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
        type="button"
      >
        <span>
          Êèê ‰∫§
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`üíµ ProFormMoney > üíµ can not input negative 1`] = `
<div>
  <form
    autocomplete="off"
    class="ant-form ant-form-vertical ant-pro-form"
  >
    <input
      style="display: none;"
      type="text"
    />
    <div
      class="ant-form-item ant-form-item-vertical"
    >
      <div
        class="ant-row ant-form-item-row"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-input-number ant-input-number-in-form-item ant-input-number-outlined ant-input-number-focused"
                style="width: 100%;"
              >
                <div
                  class="ant-input-number-handler-wrap"
                >
                  <span
                    aria-disabled="false"
                    aria-label="Increase Value"
                    class="ant-input-number-handler ant-input-number-handler-up"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="up"
                      class="anticon anticon-up ant-input-number-handler-up-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    aria-disabled="false"
                    aria-label="Decrease Value"
                    class="ant-input-number-handler ant-input-number-handler-down"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-input-number-handler-down-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-input-number-input-wrap"
                >
                  <input
                    aria-describedby="test-id"
                    aria-valuemin="0"
                    autocomplete="off"
                    class="ant-input-number-input"
                    id="amount"
                    placeholder="ËØ∑ËæìÂÖ•"
                    role="spinbutton"
                    step="1"
                    value="-55.33"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; gap: 8px; align-items: center;"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
        type="button"
      >
        <span>
          Èáç ÁΩÆ
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
        type="button"
      >
        <span>
          Êèê ‰∫§
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`üíµ ProFormMoney > üíµ moneySymbol with custom locale 1`] = `
<div>
  <form
    autocomplete="off"
    class="ant-form ant-form-vertical ant-pro-form"
  >
    <input
      style="display: none;"
      type="text"
    />
    <div
      class="ant-form-item ant-form-item-vertical"
    >
      <div
        class="ant-row ant-form-item-row"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-input-number ant-input-number-in-form-item ant-input-number-outlined ant-input-number-focused"
                style="width: 100%;"
              >
                <div
                  class="ant-input-number-handler-wrap"
                >
                  <span
                    aria-disabled="false"
                    aria-label="Increase Value"
                    class="ant-input-number-handler ant-input-number-handler-up"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="up"
                      class="anticon anticon-up ant-input-number-handler-up-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    aria-disabled="false"
                    aria-label="Decrease Value"
                    class="ant-input-number-handler ant-input-number-handler-down"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-input-number-handler-down-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-input-number-input-wrap"
                >
                  <input
                    aria-describedby="test-id"
                    aria-valuenow="44.33"
                    autocomplete="off"
                    class="ant-input-number-input"
                    id="amount"
                    placeholder="Please enter"
                    role="spinbutton"
                    step="1"
                    value="$ 44.33"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; gap: 8px; align-items: center;"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
        type="button"
      >
        <span>
          Èáç ÁΩÆ
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
        type="button"
      >
        <span>
          Êèê ‰∫§
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`üíµ ProFormMoney > üíµ moneySymbol with custom symbol 1`] = `
<div>
  <form
    autocomplete="off"
    class="ant-form ant-form-vertical ant-pro-form"
  >
    <input
      style="display: none;"
      type="text"
    />
    <div
      class="ant-form-item ant-form-item-vertical"
    >
      <div
        class="ant-row ant-form-item-row"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-input-number ant-input-number-in-form-item ant-input-number-outlined ant-input-number-focused"
                style="width: 100%;"
              >
                <div
                  class="ant-input-number-handler-wrap"
                >
                  <span
                    aria-disabled="false"
                    aria-label="Increase Value"
                    class="ant-input-number-handler ant-input-number-handler-up"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="up"
                      class="anticon anticon-up ant-input-number-handler-up-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    aria-disabled="false"
                    aria-label="Decrease Value"
                    class="ant-input-number-handler ant-input-number-handler-down"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-input-number-handler-down-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-input-number-input-wrap"
                >
                  <input
                    aria-describedby="test-id"
                    aria-valuenow="44.33"
                    autocomplete="off"
                    class="ant-input-number-input"
                    id="amount"
                    placeholder="ËØ∑ËæìÂÖ•"
                    role="spinbutton"
                    step="1"
                    value="üí∞ 44.33"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; gap: 8px; align-items: center;"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
        type="button"
      >
        <span>
          Èáç ÁΩÆ
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
        type="button"
      >
        <span>
          Êèê ‰∫§
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`üíµ ProFormMoney > üíµ moneySymbol with global locale 1`] = `
<div>
  <form
    autocomplete="off"
    class="ant-form ant-form-vertical ant-pro-form"
  >
    <input
      style="display: none;"
      type="text"
    />
    <div
      class="ant-form-item ant-form-item-vertical"
    >
      <div
        class="ant-row ant-form-item-row"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-input-number ant-input-number-in-form-item ant-input-number-outlined ant-input-number-focused"
                style="width: 100%;"
              >
                <div
                  class="ant-input-number-handler-wrap"
                >
                  <span
                    aria-disabled="false"
                    aria-label="Increase Value"
                    class="ant-input-number-handler ant-input-number-handler-up"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="up"
                      class="anticon anticon-up ant-input-number-handler-up-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    aria-disabled="false"
                    aria-label="Decrease Value"
                    class="ant-input-number-handler ant-input-number-handler-down"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-input-number-handler-down-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-input-number-input-wrap"
                >
                  <input
                    aria-describedby="test-id"
                    aria-valuenow="44.33"
                    autocomplete="off"
                    class="ant-input-number-input"
                    id="amount"
                    placeholder="Please enter"
                    role="spinbutton"
                    step="1"
                    value="¬£ 44.33"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; gap: 8px; align-items: center;"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
        type="button"
      >
        <span>
          Reset
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
        type="button"
      >
        <span>
          Submit
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`üíµ ProFormMoney > üíµ update money precision when init 1`] = `
<div>
  <form
    autocomplete="off"
    class="ant-form ant-form-vertical ant-pro-form"
  >
    <input
      style="display: none;"
      type="text"
    />
    <div
      class="ant-form-item ant-form-item-vertical"
    >
      <div
        class="ant-row ant-form-item-row"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-input-number ant-input-number-in-form-item ant-input-number-outlined ant-input-number-focused"
                style="width: 100%;"
              >
                <div
                  class="ant-input-number-handler-wrap"
                >
                  <span
                    aria-disabled="false"
                    aria-label="Increase Value"
                    class="ant-input-number-handler ant-input-number-handler-up"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="up"
                      class="anticon anticon-up ant-input-number-handler-up-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    aria-disabled="false"
                    aria-label="Decrease Value"
                    class="ant-input-number-handler ant-input-number-handler-down"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-input-number-handler-down-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-input-number-input-wrap"
                >
                  <input
                    aria-describedby="test-id"
                    aria-valuenow="444444444.3333333"
                    autocomplete="off"
                    class="ant-input-number-input"
                    id="amount"
                    placeholder="ËØ∑ËæìÂÖ•"
                    role="spinbutton"
                    step="1"
                    value="üí∞ 444,444,444.33"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; gap: 8px; align-items: center;"
    >
      <button
        class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
        type="button"
      >
        <span>
          Èáç ÁΩÆ
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
        type="button"
      >
        <span>
          Êèê ‰∫§
        </span>
      </button>
    </div>
  </form>
</div>
`;
